File: ios/RNDsbluetooth.m

#import <Foundation/Foundation.h>
#import <React/RCTBridgeModule.h>
#import <React/RCTEventEmitter.h>

@interface RCT_EXTERN_MODULE(RNDsbluetooth, RCTEventEmitter)

#pragma mark - RN方法

RCT_EXTERN_METHOD(init:(NSString *) restoreStateIdentifier enableLog:(BOOL)enableLog)

RCT_EXTERN_METHOD(sdkVersion:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(enableDebug:(BOOL) debug)

RCT_EXTERN_METHOD(startScan:(NSArray<NSString *>*) filter options:(NSDictionary<NSString *, id> *) options)

RCT_EXTERN_METHOD(stopScan)

RCT_EXTERN_METHOD(connect:(NSString *) identifier mac:(NSString *) mac)

RCT_EXTERN_METHOD(disconnect:(NSString *) identifier)

RCT_EXTERN_METHOD(state:(RCTPromiseResolveBlock)resolve)

RCT_EXTERN_METHOD(bandSend:(NSInteger)command data:(id) data resole:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(bandSyncData)

RCT_EXTERN_METHOD(bandBind)

RCT_EXTERN_METHOD(bandDfu:(id)data resole:(RCTPromiseResolveBlock)resolve reject:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(destroy)

@end

File: ios/RNDSbluetooth.swift
//
//  RNDSbluetooth.swift
//  RNDsbluetooth
//
//  Created by Jaysonj on 2019/3/21.
//  Copyright © 2019 Facebook. All rights reserved.
//

import Foundation

@objc(RNDsbluetooth)
class RNDsbluetooth: RCTEventEmitter {
    var hasListeners = false
    var bleManager: RNBleManager? = nil
    
    deinit {
        bleManager?.destroy()
        bleManager = nil
        hasListeners = false
    }
    
    override func stopObserving() {
        hasListeners = false
    }
    
    override func startObserving() {
        hasListeners = true
    }
    
    override func supportedEvents() -> [String]! {
        return RNBleEvent.events
    }
    
    @objc
    override static func requiresMainQueueSetup() -> Bool {
        return true
    }
    
    @objc(init:enableLog:)
    func initManager(restoreStateIdentifier: String, enableLog: Bool) {
        bleManager = RNBleManager.init(restoreStateIdentifier: restoreStateIdentifier, enableLog: enableLog)
        bleManager?.delegate = self
    }
    
    @objc(sdkVersion:reject:)
    func sdkVersion(_ resolve: @escaping Resolve, reject: @escaping Reject) {
        bleManager?.sdkVersion(resolve, reject: reject)
    }
    
    @objc(enableDebug:)
    func enableDebug(_ debug: Bool) {
        bleManager?.enableDebug(debug)
    }
    
    @objc(startScan:options:)
    func startScan(filter: Array<String>?, options: [String: Any]?) {
        var scanFilter: Set<String> = []
        filter?.forEach({ (it) in
            scanFilter.insert(it)
        })
        bleManager?.startScan(filter: scanFilter.isEmpty ? nil : scanFilter, options: options)
    }
    
    @objc(stopScan)
    func stopScan() {
        bleManager?.stopScan()
    }
    
    @objc(connect:mac:)
    func connect(identifier: String?, mac: String?) {
       bleManager?.connect(identifier: identifier, mac: mac)
    }
    
    @objc(disconnect:)
    func disconnect(identifier: String?) {
        bleManager?.disconnect(identifier: identifier)
    }
    
    @objc(state:)
    func state(_ resolve: @escaping Resolve) {
        bleManager?.state(resolve)
    }
    
    @objc(bandSend:data:resole:reject:)
    func bandSend(_ command: Int, data: Any?, resole: @escaping Resolve, reject:  @escaping Reject){
        bleManager?.bandSend(command, data: data, resole: resole, reject: reject)
    }
    
    @objc(bandSyncData)
    func bandSyncData() {
        bleManager?.bandSyncData()
    }
    
    @objc(bandBind)
    func bandBind() {
        bleManager?.bandBind()
    }
    
    @objc(bandDfu:resole:reject:)
    func bandDfu(_ dfuFile: Any, resole: @escaping Resolve, reject:  @escaping Reject) {
        bleManager?.bandDfu(dfuFile, resole: resole, reject: reject)
    }
    
    @objc(destroy)
    public func destroy() {
        bleManager?.destroy()
        bleManager = nil
    }
}

extension RNDsbluetooth: RNBleManagerDelegate {
    func dispatchEvent(_ name: String, value: Any?) {
        if hasListeners {
            sendEvent(withName: name, body: value)
        }
    }
}
